helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -777,6 +850,38 @@ public boolean validateTopologySettingInInstanceConfig(ClusterConfig clusterConf
     return true;
   }
 
+  /**
+   * Overwrite the InstanceConfigProperties from the given InstanceConfig to this InstanceConfig.
+   * The merge is done by overwriting the properties in this InstanceConfig with the properties
+   * from the given InstanceConfig. {@link #NON_OVERWRITABLE_PROPERTIES} will not be overridden.
+   *
+   * @param overwritingInstanceConfig the InstanceConfig to override into this InstanceConfig
+   */
+  public void overwriteInstanceConfig(InstanceConfig overwritingInstanceConfig) {
+    for (InstanceConfigProperty keyEnum : InstanceConfigProperty.values()) {
+      if (NON_OVERWRITABLE_PROPERTIES.contains(keyEnum)) {
+        continue;
+      }
+
+      String key = keyEnum.toString();
+      // Add key value pair if it exists in overwritingInstanceConfig and remove it if it doesn't
+      // exist in overwritingInstanceConfig but does exist in this InstanceConfig.
+      if (overwritingInstanceConfig.getRecord().getSimpleFields().containsKey(key)) {
+        _record.setSimpleField(key, overwritingInstanceConfig.getRecord().getSimpleField(key));
+      } else if (_record.getSimpleFields().containsKey(key)) {
+        _record.getSimpleFields().remove(key);
+      } else if (overwritingInstanceConfig.getRecord().getListFields().containsKey(key)) {
+        _record.setListField(key, overwritingInstanceConfig.getRecord().getListField(key));
+      } else if (_record.getListFields().containsKey(key)) {
+        _record.getListFields().remove(key);
+      } else if (overwritingInstanceConfig.getRecord().getMapFields().containsKey(key)) {
+        _record.setMapField(key, overwritingInstanceConfig.getRecord().getMapField(key));
+      } else if (_record.getMapFields().containsKey(key)) {
+        _record.getMapFields().remove(key);
+      }
```

Comment:
Why not just addAll the defined keys for three fields.... This is too complicated.